File: remex.android/app/src/main/java/com/clindsay94/remex/service/FileConflictPolicy.kt
package com.clindsay94.remex.service

/**
 * Wire values the host sends and accepts for a filename collision (RemEx-agpn).
 *
 * Mirrors `Remex.Core.Models.FileTransferErrorCodes` and `FileConflictResolutions` VERBATIM. These
 * are protocol tokens, not prose — the whole point of the host change that introduced them is that
 * a client can branch on something that is never translated.
 */
object FileConflictCodes {
    const val DESTINATION_EXISTS = "destination_exists"
    const val DESTINATION_IS_DIFFERENT_KIND = "destination_is_different_kind"

    /**
     * The host renamed for "keep both" and the destination refused the new name (RemEx-cirk).
     *
     * **KNOWN EXPLICITLY RATHER THAN FALLING THROUGH TO THE UNKNOWN-CODE BRANCH.** Skip is the only
     * sensible answer either way, so the ACTIONS would have been the same - but the unknown branch
     * also leaves the sheet asserting the wrong cause, and this arrives at the worst moment for
     * that: after the user has already chosen Keep both.
     */
    const val RESOLVED_NAME_UNUSABLE = "resolved_name_unusable"
}

object FileConflictResolutions {
    const val REPLACE = "replace"
    const val KEEP_BOTH = "keep_both"
}

/** What the user can do about one collision. */
enum class ConflictAction {
    /** Overwrite what is there. Never offered when the destination is a different kind of thing. */
    Replace,

    /** Let the host pick the next free name. The phone never composes it. */
    KeepBoth,

    /** Do nothing for this item and carry on. Always available, and the fail-closed default. */
    Skip,
}

/**
 * Decides which answers a collision may be offered, and remembers one across a batch (RemEx-agpn).
 *
 * **PURE, AND SEPARATE FROM THE SHEET, because the decisions are where a mistake costs a file.**
 * Offering "Replace" for a collision whose replace deletes a directory tree, or letting an
 * apply-to-all Replace leak past the batch it was given for, are both data loss — and neither is
 * visible in a screenshot of the sheet.
 */
object FileConflictPolicy {

    /**
     * The actions worth offering for [errorCode] on [operation].
     *
     * @param operation one of `FileManageOperations`.
     */
    fun actionsFor(errorCode: String?, operation: String): List<ConflictAction> {
        // Not a collision at all. A sheet must not open for an ordinary failure - "Replace" is a
        // meaningless answer to "the disk is full", and offering it invites a retry that cannot work.
        if (errorCode == null) return emptyList()

        // MKDIR IS REFUSAL-ONLY, and the host says so: CreateDirectoryAsync emits a code but accepts
        // no conflictResolution, so a Replace or Keep both retry re-fails identically. Offering them
        // produces a dead-end loop where the user answers the same question forever. The code still
        // earns its place - it says WHY the mkdir failed - and Skip remains a real answer.
        if (operation == FileManageOperations.MKDIR) return listOf(ConflictAction.Skip)

        return when (errorCode) {
            // KEEP BOTH FIRST, REPLACE SECOND. Review caught the original order putting the
            // destructive answer at the top: de-weighting its emphasis is not enough, because
            // POSITION is the stronger recommendation and the top button is the one a hurried user
            // taps. The order here is the order the sheet renders.
            FileConflictCodes.DESTINATION_EXISTS ->
                listOf(ConflictAction.KeepBoth, ConflictAction.Replace, ConflictAction.Skip)

            // NO REPLACE. Here the destination is the OTHER KIND of thing, so replacing means
            // deleting a whole directory tree to make room for one file, or deleting a file to make
            // a folder. Nobody intends either from a copy and nothing undoes them. The host refuses
            // it outright, so the button would not even work - but the reason it is withheld is that
            // it should never have been offered.
            FileConflictCodes.DESTINATION_IS_DIFFERENT_KIND ->
                listOf(ConflictAction.KeepBoth, ConflictAction.Skip)


            // SKIP ALONE FOR EVERYTHING ELSE, INCLUDING RESOLVED_NAME_UNUSABLE. An explicit branch
            // for that code was written here and removed: the else already yields Skip, so no mutant
            // could kill it, and an unkillable branch reads to a later maintainer as though it
            // decides something. The code IS known to this client - the sheet gives it its own
            // explanation - but the ACTIONS genuinely are the same, and for a good reason: the host
            // already tried the rename and the destination refused the name it chose, so retrying
            // keep-both asks it to choose again from the same too-long stem, and replace was
            // declined a moment ago. Nothing is left that could work.
            //
            // A CODE THIS CLIENT DOES NOT KNOW. A newer host may send one, and guessing which
            // actions it permits is how a client offers a destructive button for a situation it does
            // not understand. Skip is the only answer that is safe against every possible meaning.
            else -> listOf(ConflictAction.Skip)
        }
    }

    /** The wire value for [action], or null when the action sends no retry at all. */
    fun resolutionFor(action: ConflictAction): String? = when (action) {
        ConflictAction.Replace -> FileConflictResolutions.REPLACE
        ConflictAction.KeepBoth -> FileConflictResolutions.KEEP_BOTH

        // SKIP SENDS NOTHING. It is not a resolution the host understands - it is the client
        // declining to retry, which is why it can never fail and why it is the safe default.
        ConflictAction.Skip -> null
    }
}

/**
 * One batch's answer to "apply to all remaining" (RemEx-agpn).
 *
 * **SCOPED TO A BATCH BY CONSTRUCTION.** A remembered Replace that outlived the operation it was
 * given for would overwrite a file in some later, unrelated copy that the user was never asked
 * about — so this is created per batch and discarded with it, rather than living on the ViewModel
 * where forgetting to clear it is a silent data-loss bug.
 */
class BatchConflictChoice {
    private var remembered: ConflictAction? = null

    /** The answer to reuse without asking, or null when the user must be asked. */
    val standingAnswer: ConflictAction? get() = remembered

    /**
     * Records [action] for the rest of this batch.
     *
     * @param applyToAll false for a one-off answer, which is the default and changes nothing.
     */
    fun remember(action: ConflictAction, applyToAll: Boolean) {
        if (applyToAll) remembered = action
    }

    /**
     * Whether [action] can stand in for a collision reporting [errorCode] on [operation].
     *
     * **A STANDING ANSWER IS STILL CHECKED AGAINST EACH COLLISION, and that is the point.** A user
     * who chose "Replace, apply to all" for a batch of ordinary collisions has not agreed to delete
     * a directory tree when the next item turns out to be a different kind of thing — a remembered
     * answer that bypassed the offer list would do exactly that, silently, on an item they never saw.
     */
    fun canApply(action: ConflictAction, errorCode: String?, operation: String): Boolean =
        action in FileConflictPolicy.actionsFor(errorCode, operation)
}

File: remex.desktop/Services/FileTransfer/FileConflictPolicy.cs
using Remex.Core.Models;

namespace Remex.Desktop.Services.FileTransfer;

/// <summary>What the person at the PC can do about one filename collision.</summary>
public enum FileConflictAction
{
    /// <summary>Overwrite what is there. Never offered when the destination is a different kind of thing.</summary>
    Replace,

    /// <summary>Let the host pick the next free name. The PC never composes it.</summary>
    KeepBoth,

    /// <summary>Leave this item alone and carry on with the rest of the paste.</summary>
    Skip,

    /// <summary>Abandon the whole paste, including the items not reached yet.</summary>
    Cancel,
}

/// <summary>
/// Decides which answers a collision may be offered on the PC side.
/// </summary>
/// <remarks>
/// <para>
/// **A C# MIRROR OF <c>FileConflictPolicy.kt</c>, DELIBERATELY, AND NOT A NEW DIALECT.** The wire
/// vocabulary is <see cref="FileTransferErrorCodes"/> and <see cref="FileConflictResolutions"/> in
/// <c>remex.core</c>; both clients read the same constants and reach the same offer lists from
/// them. A PC that offered a different set for the same code would mean the two clients disagree
/// about which answers are SAFE, and every disagreement of that kind resolves in favour of data
/// loss on one of them.
/// </para>
/// <para>
/// **PURE, AND SEPARATE FROM THE VIEW MODEL, because the decisions are where a mistake costs a
/// file.** Offering Replace for a collision whose replace deletes a directory tree is not visible
/// in a screenshot of the prompt, and nothing undoes it.
/// </para>
/// <para>
/// NO <c>mkdir</c> BRANCH, unlike the Kotlin. The phone's sheet is reached from every manage
/// operation; this prompt is reached only from paste, so a branch for mkdir here would be a rule
/// no call site can exercise and no mutant can kill — the shape this repo's own sweeps flag
/// (RemEx-thwlr). <c>MakeDirectoryRemoteAsync</c> keeps throwing, which is refusal-only by another
/// route.
/// </para>
/// </remarks>
public static class FileConflictPolicy
{
    /// <summary>The actions worth offering for <paramref name="errorCode"/>, in render order.</summary>
    /// <remarks>
    /// KEEP BOTH FIRST, REPLACE SECOND, matching the phone. Review caught the Android ordering
    /// putting the destructive answer at the top, and position is a stronger recommendation than
    /// any amount of de-emphasis: the first button is the one a hurried user presses.
    /// </remarks>
    public static IReadOnlyList<FileConflictAction> ActionsFor(string? errorCode)
    {
        // Not a collision at all. An ordinary failure — the disk filled up, the folder went
        // read-only — has no answer a prompt could collect, and offering "Replace" for one invites
        // a retry that cannot work.
        if (string.IsNullOrWhiteSpace(errorCode))
            return [];

        return errorCode switch
        {
            FileTransferErrorCodes.DestinationExists =>
                [FileConflictAction.KeepBoth, FileConflictAction.Replace, FileConflictAction.Skip, FileConflictAction.Cancel],

            // KEEP BOTH AND SKIP, NEVER REPLACE. The name is taken, so asking again genuinely works
            // — the host re-lists the directory and picks the next free name. But Replace answers
            // the ORIGINAL request, overwriting the destination the user first named while the
            // prompt is showing the sibling the host invented. A user who chose keep-both precisely
            // to preserve that file would destroy it by answering a question about a different one.
            FileTransferErrorCodes.ResolvedNameTaken =>
                [FileConflictAction.KeepBoth, FileConflictAction.Skip, FileConflictAction.Cancel],

            // NO REPLACE. The destination is the OTHER KIND of thing, so replacing means deleting a
            // whole directory tree to make room for one file, or deleting a file to make a folder.
            // Nobody intends either from a paste and nothing undoes them; the host refuses it
            // outright, so the button would not even work — but the reason it is withheld is that
            // it should never have been offered.
            FileTransferErrorCodes.DestinationIsDifferentKind =>
                [FileConflictAction.KeepBoth, FileConflictAction.Skip, FileConflictAction.Cancel],

            // SKIP OR CANCEL FOR EVERYTHING ELSE, INCLUDING resolved_name_unusable. That code gets
            // its own SENTENCE (see FileConflictPrompt) but not its own branch, because the actions
            // genuinely are the same and an explicit case would be unkillable by any mutant — the
            // exact reasoning the Kotlin records for deleting its own copy of it. The host already
            // tried the rename and the destination refused the name it chose, so asking it to
            // choose again from the same too-long stem cannot work, and Replace was declined a
            // moment ago.
            //
            // A CODE THIS BUILD DOES NOT KNOW lands here too. Guessing which actions a newer host
            // permits is how a client offers a destructive button for a situation it does not
            // understand.
            _ => [FileConflictAction.Skip, FileConflictAction.Cancel],
        };
    }

    /// <summary>The wire value for <paramref name="action"/>, or null when it sends no retry at all.</summary>
    public static string? ResolutionFor(FileConflictAction action) => action switch
    {
        FileConflictAction.Replace => FileConflictResolutions.Replace,
        FileConflictAction.KeepBoth => FileConflictResolutions.KeepBoth,

        // SKIP AND CANCEL SEND NOTHING. Neither is a resolution the host understands — they are the
        // PC declining to retry, which is why neither can fail and why Skip is the safe default.
        _ => null,
    };
}
